libtransact/src/state/merkle.rs

```diff
@@ -202,23 +210,25 @@ impl MerkleRadixTree {
         } else if change_log.successors.is_empty() {
             // deleting the tip of a trie lineage
 
-            let (deletion_candidates, duplicates) =
-                MerkleRadixTree::remove_duplicate_hashes(&mut db_writer, change_log.additions)?;
+            let (deletion_candidates, duplicates) = MerkleRadixTree::remove_duplicate_hashes(
+                db_writer.as_reader(),
+                change_log.additions,
+            )?;
 
             for hash in &deletion_candidates {
                 let hash_hex = ::hex::encode(hash);
-                delete_ignore_missing(&mut db_writer, hash_hex.as_bytes())?
+                delete_ignore_missing(db_writer.as_mut(), hash_hex.as_bytes())?
```

Comment:
This is interesting.  What required this change?

Reply:
I'm not sure why, but when using `Box` or `Arc` for traits as objects I can't pass references using `&`. 
If I were to leave it `delete_ignore_missing(&mut db_writer, hash_hex.as_bytes())?` 
I get the following compiler error:

```
error[E0277]: the trait bound `std::boxed::Box<dyn database::DatabaseWriter>: database::DatabaseWriter` is not satisfied
   --> libtransact/src/state/merkle.rs:216:39
    |
216 |                 delete_ignore_missing(&mut db_writer, hash_hex.as_bytes())?
    |                                       ^^^^^^^^^^^^^^ the trait `database::DatabaseWriter` is not implemented for `std::boxed::Box<dyn database::DatabaseWriter>`
    |
    = note: required for the cast to the object type `dyn database::DatabaseWriter`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: Could not compile `transact`.

```

Reply:
I think you can do just do `&mut *db_writer`

See https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5a959935efc6601504c50aed24c432cb as an example.

Reply:
Updated